### ISSUE TITLE
Palettes data structures

### ISSUE THREAD

**Issue by rcannood** [funkyheatmap/funkyheatmap#20]
At the moment, the "minimal" example uses 6 palettes ( https://github.com/funkyheatmap/funkyheatmap.github.io/blob/main/examples/minimal/_script.R#L45-L66 ), namely bar, circle, funkyrect, rect, black6white4 and pie.

There are three ways you can define a palette:

1. not providing a value for the palette, in which the package will choose a palette for you
2. use one of the colorbrewer palettes by providing the name of the palette:
  ```r
  palettes <- list(
    text = "Greys",
    bar = "Blues",
    ...,
    pie = "Set2"
  )
  ```
3. manually define the hex codes:
  ```r
  palettes <- list(
    text = c("#000000", "#020202", "#050505", ...),
    ...
    pie = c(foo = "#8DD3C7", bar = "#FFFFB3", zing = "#BEBADA")
  )
  ```

---

However, @mxposed mentioned the latter causes issues in JS because it's hard to recover the ordering of the keys in order to render the legend. @mxposed Why is this the case? What is the problem with using `Object.keys()`?

```js
> x = {foo: "#8DD3C7", zing: "#FFFFB3", bar: "#BEBADA"}
{ foo: '#8DD3C7', zing: '#FFFFB3', bar: '#BEBADA' }
> Object.keys(x)
[ 'foo', 'zing', 'bar' ]
```

(I must be missing something here)

---

Regardless, the current format also doesn't allow to tune how the legends are rendered - e.g. which title is assigned to the palette. What about supporting the following format:

```yaml
palettes:
  # palette is not defined, so funkyheatmap will automatically choose one for the user
  text: null
  
  # colorbrewer palette is used
  bar: Blues
  
  # vector of hexcodes notation
  circle: ["#000000", "#020202", "#050505", ...]
  
  # extended notation
  funkyrect:
    title: "FunkyRect"
    colors: ["#000000", "#020202", "#050505", ...]
    legend: false # in case you don't want to render a legend for this palette
  
  # extended notation for pie
  pie:
    title: "Pie"
    colors: ["#8DD3C7", "#FFFFB3", "#BEBADA"]
    keys: [foo, ing, bar]
```

@LouiseDck @mxposed @artuurC @lazappi WDYT?

**Comment by mxposed** [comment]
Ok, my concern was that the order of Object.keys is not guaranteed to be the same as when you create, so the correspondence between colors and items in data could be broken.
However, it appears that my understanding was outdated [e.g. from here](https://stackoverflow.com/questions/30076219/does-es6-introduce-a-well-defined-order-of-enumeration-for-object-properties) so I am ok to switch to Object.

I like the Object-based specification with legend and legend title

**Comment by rcannood** [comment]
I see, I didn't know about the history of `Object.keys` not guaranteeing property order!

In that case, let's stick with:

```yaml
palettes:
  # palette is not defined, so funkyheatmap will automatically choose one for the user
  text: null
  
  # colorbrewer palette is used
  funkyrect: Blues
  
  # vector of hexcodes notation
  circle: ["#000000", "#020202", "#050505", ...]
  
  # dictionary notation for pie
  pie:
    foo: "#8DD3C7"
    ing: "#FFFFB3"
    bar: "#BEBADA"
```

@mxposed Is that ok with you?

---

Regarding the "extended" notation `{title: ..., colors: ..., legend: ...}`; I was tinkering with the improving the way legends are rendered ([funkyheatmap/funkyheatmap@refactor_legends](https://github.com/funkyheatmap/funkyheatmap/tree/refactor_legends)), and I'm thinking that trying to shoe-horn this into the palettes data structure using the 'extended notation' may not be a good fit.

For one, if we allow for using separate values for size and color (#9), we may also want to plot two different legends for representing both the size and the color scales separately. If the standard legends generated by funkyheatmap are not doing what you expect it to do, maybe it would be better to allow tuning the legends by having a _separate_ parameter, e.g.:

```yaml
palettes:
  # palette is not defined, so funkyheatmap will automatically choose one for the user
  text: null
  
  # colorbrewer palette is used
  funkyrect: Blues
  
  # vector of hexcodes notation
  circle: ["#000000", "#020202", "#050505", ...]
  
  # dictionary notation for pie
  pie:
    foo: "#8DD3C7"
    ing: "#FFFFB3"
    bar: "#BEBADA"

legend:
  # This will draw a legend for the 'circle' palette using circles
  - title: Circle data
    geom: circle
    palette: circle

  # This will draw a custom legend for visualising the funkyrect rank-based palette
  - title: Funkyrect rank
    geom: funkyrect
    palette: funkyrect
    keys: ["1", "", "", "", "", "n"]
    size: 1 # fixed size, can also be an array of values

  # This will draw a custom legend for visualising the funkyrect size
  - title: Funkyrect value
    geom: funkyrect
    color: lightgray # fixed colour, can also be an array of values
```

I expect that the legends specified above should yield the following results:

![Screenshot from 2023-07-07 09-04-02](https://github.com/funkyheatmap/funkyheatmap/assets/553642/61ec6db8-d12e-4d15-82f5-debbfbe9e33e)

**Comment by rcannood** [comment]
@mxposed @LouiseDck @artuurC I've implemented a first version of being able to configure the legends in #23. It consists of three main parts:

* verify_palettes.R: Check whether what is passed in the palettes list makes sense, and fill in sensible defaults if no settings are defined
* create_legends.R: Helper functions for creating ggplots for each of the desired legends. In essence it first creates the same data structures as are being used by the funky_heatmap, and then runs `compose_ggplot` to create the plot.
* funky_heatmap.R: documentation regarding the structure of the legends has been added. patchwork is used to combine all legends on one row, and stack the main plot on top of the legends.

On the minimum reproducible example, not specifying any legends results in:

![plot](https://github.com/funkyheatmap/funkyheatmap/assets/553642/847d801e-48d6-4c0f-b771-3bef8e63c964)

---

However, defining legends as

```r

legends <-
  list(
    list(
      title = "Rects",
      palette = "rect",
      enabled = TRUE,
      geom = "rect",
      size = 1,
      labels = c("0", "", "", "0.5", "", "", "1")
    ),
    list(
      title = "Funkyrect rank",
      palette = "funkyrect",
      enabled = TRUE,
      geom = "funkyrect",
      size = 1,
      labels = c("1", "", "", "", "", "n")
    ),
    list(
      title = "Funkyrect value",
      enabled = TRUE,
      geom = "funkyrect",
      color = "gray80"
    ),
    list(
      title = "Circles",
      palette = "circle",
      enabled = TRUE,
      geom = "circle"
    )
  )
```

Results in the following figure:

![plot](https://github.com/funkyheatmap/funkyheatmap/assets/553642/ecbd0046-e835-4c21-9857-6ed313cf0735)

---

Some minor tweaking is still in order. Also, there is currently no way for rendering labels on top of the geoms in the legend, I'll have to add that later.
